plugins/janus_nosip.c

```diff
@@ -0,0 +1,2213 @@
+	json_t *info = json_object();
+	json_object_set_new(info, "info", session->info ? json_string(session->info) : NULL);
+	if(session->sdp) {
+		json_object_set_new(info, "srtp-required", json_string(session->media.require_srtp ? "yes" : "no"));
+		json_object_set_new(info, "sdes-local", json_string(session->media.has_srtp_local ? "yes" : "no"));
+		json_object_set_new(info, "sdes-remote", json_string(session->media.has_srtp_remote ? "yes" : "no"));
+	}
+	if(session->arc || session->vrc || session->arc_peer || session->vrc_peer) {
+		json_t *recording = json_object();
+		if(session->arc && session->arc->filename)
+			json_object_set_new(recording, "audio", json_string(session->arc->filename));
+		if(session->vrc && session->vrc->filename)
+			json_object_set_new(recording, "video", json_string(session->vrc->filename));
+		if(session->arc_peer && session->arc_peer->filename)
+			json_object_set_new(recording, "audio-peer", json_string(session->arc_peer->filename));
+		if(session->vrc_peer && session->vrc_peer->filename)
+			json_object_set_new(recording, "video-peer", json_string(session->vrc_peer->filename));
+		json_object_set_new(info, "recording", recording);
+	}
+	json_object_set_new(info, "destroyed", json_integer(session->destroyed));
+	return info;
+}
+
+struct janus_plugin_result *janus_nosip_handle_message(janus_plugin_session *handle, char *transaction, json_t *message, json_t *jsep) {
+	if(g_atomic_int_get(&stopping) || !g_atomic_int_get(&initialized))
+		return janus_plugin_result_new(JANUS_PLUGIN_ERROR, g_atomic_int_get(&stopping) ? "Shutting down" : "Plugin not initialized", NULL);
+	janus_nosip_message *msg = g_malloc0(sizeof(janus_nosip_message));
+	msg->handle = handle;
+	msg->transaction = transaction;
+	msg->message = message;
+	msg->jsep = jsep;
+	g_async_queue_push(messages, msg);
+
+	/* All the requests to this plugin are handled asynchronously */
+	return janus_plugin_result_new(JANUS_PLUGIN_OK_WAIT, NULL, NULL);
+}
+
+void janus_nosip_setup_media(janus_plugin_session *handle) {
+	JANUS_LOG(LOG_INFO, "WebRTC media is now available\n");
+	if(g_atomic_int_get(&stopping) || !g_atomic_int_get(&initialized))
+		return;
+	janus_nosip_session *session = (janus_nosip_session *)handle->plugin_handle;
+	if(!session) {
+		JANUS_LOG(LOG_ERR, "No session associated with this handle...\n");
+		return;
+	}
+	if(session->destroyed)
+		return;
+	g_atomic_int_set(&session->hangingup, 0);
+}
+
+void janus_nosip_incoming_rtp(janus_plugin_session *handle, int video, char *buf, int len) {
+	if(handle == NULL || handle->stopped || g_atomic_int_get(&stopping) || !g_atomic_int_get(&initialized))
+		return;
+	if(gateway) {
+		/* Honour the audio/video active flags */
+		janus_nosip_session *session = (janus_nosip_session *)handle->plugin_handle;
+		if(!session || session->destroyed) {
+			JANUS_LOG(LOG_ERR, "No session associated with this handle...\n");
+			return;
+		}
+		/* Forward to our NoSIP peer */
+		if(video) {
```

Comment:
Good point, which probably applies to other plugins as well. I'll keep that in mind.